File: Java/Structural/FlyweightPattern.java
package Structural;

import java.util.HashMap;
import java.util.Map;

// Flyweight
interface Character {
    void print();
}

// Concrete Flyweight
class A implements Character {
    @Override
    public void print() {
        System.out.print("A");
    }
}

class B implements Character {
    @Override
    public void print() {
        System.out.print("B");
    }
}

// Flyweight Factory
class CharacterFactory {
    private static final Map<String, Character> characters = new HashMap<>();
    
    public static Character getCharacter(String key) {
        if (!characters.containsKey(key)) {
            if (key.equals("A")) {
                characters.put("A", new A());
            } else if (key.equals("B")) {
                characters.put("B", new B());
            }
        }
        return characters.get(key);
    }
}

File: Python/Structural/FlyweightPattern.py
# Flyweight
class Character:
    def print(self):
        pass

# Concrete Flyweights
class A(Character):
    def print(self):
        print("A", end="")

class B(Character):
    def print(self):
        print("B", end="")

# Flyweight Factory
class CharacterFactory:
    _characters = {}

    @staticmethod
    def get_character(key):
        if key not in CharacterFactory._characters:
            if key == "A":
                CharacterFactory._characters[key] = A()
            elif key == "B":
                CharacterFactory._characters[key] = B()
        return CharacterFactory._characters[key]

# Client usage
char_a = CharacterFactory.get_character("A")
char_b = CharacterFactory.get_character("B")

char_a.print()  # A
char_b.print()  # B
# Flyweight
class Character:
    def print(self):
        pass

# Concrete Flyweights
class A(Character):
    def print(self):
        print("A", end="")

class B(Character):
    def print(self):
        print("B", end="")

# Flyweight Factory
class CharacterFactory:
    _characters = {}

    @staticmethod
    def get_character(key):
        if key not in CharacterFactory._characters:
            if key == "A":
                CharacterFactory._characters[key] = A()
            elif key == "B":
                CharacterFactory._characters[key] = B()
        return CharacterFactory._characters[key]

# Client usage
char_a = CharacterFactory.get_character("A")
char_b = CharacterFactory.get_character("B")

char_a.print()  # A
char_b.print()  # B
